pkg/querier-rf1/wal/querier.go

```diff
@@ -140,6 +240,94 @@ func (q *Querier) matchingChunks(ctx context.Context, tenantID string, from, thr
 	return lazyChunks, nil
 }
 
+func (q *Querier) matchingSeries(ctx context.Context, tenantID string, from, through int64, matchers ...*labels.Matcher) ([]labels.Labels, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "matchingSeries")
+	defer sp.Finish()
+	// todo support sharding
+	var (
+		series = make([]labels.Labels, 0, 1024)
+		found  = make(map[uint64]struct{})
+		mtx    sync.Mutex
+	)
+
+	err := q.forSeries(ctx, &metastorepb.ListBlocksForQueryRequest{
+		TenantId:  tenantID,
+		StartTime: from,
+		EndTime:   through,
+	}, func(id string, lbs *labels.ScratchBuilder, chk *chunks.Meta) error {
+		mtx.Lock()
+		defer mtx.Unlock()
+		lbls := lbs.Labels()
+		lblHash, _ := lbls.HashWithoutLabels([]byte(index.TenantLabel))
+		if _, ok := found[lblHash]; ok {
+			return nil
+		}
+		found[lblHash] = struct{}{}
+		// Remove tenant labels
+		lbs.Sort()
+		newLbs := lbs.Labels()
+		j := 0
+		for _, l := range newLbs {
+			if l.Name != index.TenantLabel {
+				newLbs[j] = l
+				j++
+			}
+		}
+		newLbs = newLbs[:j]
+		series = append(series, newLbs)
+		return nil
+	}, matchers...)
+	if err != nil {
+		return nil, err
+	}
+	if sp != nil {
+		sp.LogKV("matchedSeries", len(series))
+	}
+	sort.Slice(series, func(i, j int) bool {
+		return labels.Compare(series[i], series[j]) < 0
+	})
+	return series, nil
+}
+
+func (q *Querier) matchingLabels(ctx context.Context, tenantID string, from, through int64, matchers ...*labels.Matcher) ([]string, error) {
```

Comment:
I had a branch before PTO that was doing [matching labels names ](https://github.com/grafana/loki/compare/main...cyriltovena:loki:wal-read-labels?expand=1#diff-4cc270adde4ed483cfb3b5d175b57b3a0181d24685492b9a8d511aa227c65efcR114-R134 ) leveraging faster index function 

[`func (r *Reader) LabelNames(ctx context.Context, matchers ...*labels.Matcher) ([]string, error)`](https://github.com/grafana/loki/compare/main...cyriltovena:loki:wal-read-labels?expand=1#diff-06efd3e48ca000d377ab7db2c3f110e352734804d630a8b0f544f95a78cf2576R1756-R1787)

It was untested though but I think it's worth looking at.